README.md
# elisp-refs
[![Coverage Status](https://coveralls.io/repos/github/Wilfred/elisp-refs/badge.svg?branch=master)](https://coveralls.io/github/Wilfred/elisp-refs?branch=master)
[![MELPA](http://melpa.org/packages/elisp-refs-badge.svg)](http://melpa.org/#/elisp-refs)

elisp-refs is an intelligent code search for Emacs lisp.

It can find references to functions, macros or variables. Unlike a
dumb text search, elisp-refs actually parses the code, so it's never
confused by comments or variables with the same name as functions.

![screenshot](refs_screenshot.png)

This is particularly useful for finding all the places a function is
used, or finding examples of usage.

Interested readers may enjoy my blog post:
[Searching A Million Lines Of Lisp](http://www.wilfred.me.uk/blog/2016/09/30/searching-a-million-lines-of-lisp/).

## Installation

Install from MELPA (recommended) or just add elisp-refs to your `load-path`.

## Commands available

* `elisp-refs-function` (find function calls)
* `elisp-refs-macro` (find macro calls)
* `elisp-refs-variable` (find variable references)
* `elisp-refs-special` (find special form calls)
* `elisp-refs-symbol` (find all references to a symbol)

These command search all the files currently loaded in your Emacs
instance.

If called with a prefix, you can limit search results to specific
directories. For example:

`C-u M-x elisp-refs-macro RET pcase RET ~/.emacs.d/elpa/magit-20160927.510 RET`

will search for uses of `pcase` in magit:

![filtering screenshot](refs_filtered.png)

## Semantic analysis

elisp-refs has *street smarts*: given `(defun foo (bar) (baz))`, it
understands that `bar` is a variable and `baz` is a function.

elisp-refs understands the following forms:

* `defun` `defsubst` `defmacro` `cl-defun`
* `lambda`
* `let` `let*`
* `funcall` `apply`
* sharp quoted expressions (e.g. `#'some-func`)

## Limitations

elisp-refs understands elisp special forms, and a few common
macros. However, it **cannot understand arbitrary macros**.

Therefore elisp-refs will assume that `(other-macro (foo bar))` is a
function call to `foo`. If this is incorrect, you may wish to use the
command `elisp-refs-symbol` to find all references to the `foo` symbol.

If `other-macro` is a common macro, please consider submitting a patch
to `elisp-refs--function-p` to make elisp-refs smarter.

elisp-refs also does not support **indirect calls**.

``` emacs-lisp
;; Since we do a simple syntax tree walk, this isn't treated as a
;; call to foo.
(let ((x (symbol-function 'foo)))
  (funcall x))

;; Similarly, indirect function calls are not treated as
;; function calls.
(defun call-func (x)
  (funcall x))
(call-func 'foo)

;; However, if you use sharp quoting, elisp-refs knows it's a function
reference!
(call-func #'foo)
```

## Running tests

You can run the tests with:

```
$ cask install
$ cask exec ert-runner
```

## Performance

elisp-refs is CPU-intensive elisp and has been carefully optimised. You
can run the benchmark script with:

```
$ cask install
$ ./bench.sh
```

New features are carefully measured to ensure performance does not get
worse.

See elisp-refs-bench.el for more details.

## Alternative Projects

**xref-find-references**: This command is included in Emacs 25.1, but
it's based on a text search. It is confused by comments and strings,
and cannot distinguish between functions and variables.

xrefs-find-references is also line oriented, so it does not show the
whole sexp that matched your search. Since it requires text files,
it doesn't search built-in .el.gz files.

**TAGS**: It is possible to record function references in TAGS
files. Whilst [universal-ctags](https://github.com/universal-ctags/ctags) (formerly
known as exuberant-ctags) does provide the ability to find references,
it is not supported in its lisp parser.

etags, the TAGS implementation shipped with Emacs, cannot find
references (to my knowledge).

**[el-search](https://elpa.gnu.org/packages/el-search.html)** allows
you to search for arbitrary forms in elisp files. It's slower, but a
much more general tool. Its design greatly influenced elisp-refs.

**[elisp-slime-nav](https://github.com/purcell/elisp-slime-nav)**
finds definitions, not references. It's a great complementary tool.


CHANGELOG.md
## v1.6

## v1.5 (tagged 9 March 2023)

Fixed Emacs 29, which would previously crash or not find symbols.

## v1.4 (tagged 28 January 2022)

Fixed an issue where elisp-refs would kill its own results buffer if
no results were found.

Fixed an issue when running elisp-refs on Emacs master, which renamed
`format-proper-list-p` to `proper-list-p` (and moved it to subr.el).

Fixed an issue where the results buffer showed the wrong lines if
comments contained parentheses.

Fixed an issue where arguments in `declare-function` were confused
with a function call.

Fixed an error when pressing RET on a line that didn't have a search
result.

## v1.3

* Refs buffers now have names of the form `*refs: foo*`.
* Fixed an issue with a dependency on a loop.el version that doesn't
  exist.

## v1.2

* You can now filter search results to a directory. This is useful
  when working on large elisp codebases, and it's faster too.
* Results buffers now include a link to describe the thing being
  searched for.

## v1.1

* Rebranded to elisp-refs.
* Commands are now autoloaded.
* Added examples to the readme of cases that we can't detect.
* Sharp-quoted function references are now highlighted with context.
* Give more feedback on first run, when we're decompressing .el.gz
  files.
* Searches now default to the symbol at point.

## v1.0 

Initial release.



test/unit-test.el
(require 'ert)
(require 'elisp-refs)

;; For Travis CI is recursing more deeply, meaning we hit recursion
;; limits. I suspect this is due to undercover collecting coverage
;; metrics.
(when (getenv "TRAVIS")
  (message "Updating recursion limits from: max-specpdl-size: %s max-lisp-eval-depth: %s "
           max-specpdl-size
           max-lisp-eval-depth)
  (setq max-specpdl-size 2500)
  (setq max-lisp-eval-depth 1000))

(defmacro with-temp-backed-buffer (contents &rest body)
  "Create a temporary file with CONTENTS, and evaluate BODY
whilst visiting that file."
  (let ((filename-sym (make-symbol "filename"))
        (buf-sym (make-symbol "buf")))
    `(let* ((,filename-sym (make-temp-file "with-temp-buffer-and-file"))
            (,buf-sym (find-file-noselect ,filename-sym)))
       (unwind-protect
           (with-current-buffer ,buf-sym
             (insert ,contents)
             (shut-up (save-buffer))
             ,@body)
         (kill-buffer ,buf-sym)
         (delete-file ,filename-sym)))))

(ert-deftest elisp-refs--format-int ()
  "Ensure we format thousands correctly in numbers."
  (should (equal (elisp-refs--format-int 123) "123"))
  (should (equal (elisp-refs--format-int -123) "-123"))
  (should (equal (elisp-refs--format-int 1234) "1,234"))
  (should (equal (elisp-refs--format-int -1234) "-1,234"))
  (should (equal (elisp-refs--format-int 1234567) "1,234,567")))

(ert-deftest elisp-refs--pluralize ()
  (should (equal (elisp-refs--pluralize 0 "thing") "0 things"))
  (should (equal (elisp-refs--pluralize 1 "thing") "1 thing"))
  (should (equal (elisp-refs--pluralize 2 "thing") "2 things"))
  (should (equal (elisp-refs--pluralize 1001 "thing") "1,001 things")))

(ert-deftest elisp-refs--unindent-split-properties ()
  "Ensure we can still unindent when properties are split
into separate region. Regression test for a very subtle bug."
  (let ((s #("e.\n" 0 2 (elisp-refs-start-pos 0) 2 3 (elisp-refs-start-pos 0))))
    (elisp-refs--unindent-rigidly s)))

(ert-deftest elisp-refs--sexp-positions ()
  "Ensure we calculate positions correctly when we're considering
the whole buffer."
  (with-temp-backed-buffer
   "(while list (setq len 1))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (sexp-positions
           (elisp-refs--sexp-positions elisp-refs-buf (point-min) (point-max))))
     (should
      (equal sexp-positions (list '(1 26)))))))

(ert-deftest elisp-refs--sexp-positions-comments ()
  "Ensure we handle comments correctly when calculating sexp positions."
  (with-temp-backed-buffer
   "(while list
  ;; take the head of LIST
  (setq len 1))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (sexp-positions
           (elisp-refs--sexp-positions elisp-refs-buf (1+ (point-min)) (1- (point-max)))))
     ;; The position of the setq should take into account the comment.
     (should
      (equal (nth 2 sexp-positions) '(42 54))))))

(ert-deftest elisp-refs--find-calls-basic ()
  "Find simple function calls."
  (with-temp-backed-buffer
   "(foo)"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(foo) 1 6)))))))

(ert-deftest elisp-refs--find-calls-sharp-quote ()
  "Find function references using sharp quotes."
  (with-temp-backed-buffer
   "(bar #'foo)"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(bar #'foo) 1 12)))))))

(ert-deftest elisp-refs--find-calls-in-lambda ()
  "Find function calls in lambda expressions."
  (with-temp-backed-buffer
   "(lambda (foo) (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(foo) 15 20)))))))

(ert-deftest elisp-refs--find-calls-in-backquote ()
  "Find function calls in backquotes.
Useful for finding references in macros, but this is primarily a
regression test for bugs where we miscalculated position with
backquote forms."
  (with-temp-backed-buffer
   "(baz `(biz (foo 1)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(foo 1) 12 19)))))))

(ert-deftest elisp-refs--find-macros-improper-list ()
  "We shouldn't crash if the source code contains improper lists."
  (with-temp-backed-buffer
   "(destructuring-bind (start . end) region\n  (when foo\n    (bar)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'when
                                            #'elisp-refs--macro-p)))
     (should
      (equal calls (list (list '(when foo (bar)) 44 64)))))))

(ert-deftest elisp-refs--find-calls-nested ()
  "Find nested function calls."
  (with-temp-backed-buffer
   "(baz (bar (foo)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(foo) 11 16)))))))

(ert-deftest elisp-refs--find-calls-funcall ()
  "Find calls that use funcall."
  (with-temp-backed-buffer
   "(funcall 'foo)"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(funcall 'foo) 1 15)))))))

(ert-deftest elisp-refs--find-calls-apply ()
  "Find calls that use apply."
  (with-temp-backed-buffer
   "(apply 'foo '(1 2))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal calls (list (list '(apply 'foo '(1 2)) 1 20)))))))

(ert-deftest elisp-refs--find-calls-params ()
  "Function or macro parameters should not be considered function calls."
  (with-temp-backed-buffer
   "(defun bar (foo)) (defsubst bar (foo)) (defmacro bar (foo)) (cl-defun bar (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-calls-let-without-assignment ()
  "We shouldn't confuse let declarations with function calls."
  (with-temp-backed-buffer
   "(let (foo)) (let* (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-calls-let-with-assignment ()
  "We shouldn't confuse let assignments with function calls."
  (with-temp-backed-buffer
   "(let ((foo nil))) (let* ((foo nil)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-calls-let-with-assignment-call ()
  "We should find function calls in let assignments."
  ;; TODO: actually check positions, this is error-prone.
  (with-temp-backed-buffer
   "(let ((bar (foo)))) (let* ((bar (foo))))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should
      (equal (length calls) 2)))))

(ert-deftest elisp-refs--find-calls-let-body ()
  "We should find function calls in let body."
  (with-temp-backed-buffer
   "(let (bar) (foo)) (let* (bar) (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--function-p)))
     (should (equal (length calls) 2)))))

(ert-deftest elisp-refs--find-macros-basic ()
  "Find simple function calls."
  (with-temp-backed-buffer
   "(foo)"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should
      (equal calls (list (list '(foo) 1 6)))))))

(ert-deftest elisp-refs--find-macros-in-lambda ()
  "Find macros calls in lambda expressions."
  (with-temp-backed-buffer
   "(lambda (foo) (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should
      (equal calls (list (list '(foo) 15 20)))))))

(ert-deftest elisp-refs--find-macros-params ()
  "Find simple function calls."
  (with-temp-backed-buffer
   "(defun bar (foo)) (defsubst bar (foo)) (defmacro bar (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-macros-let-without-assignment ()
  "We shouldn't confuse let declarations with macro calls."
  (with-temp-backed-buffer
   "(let (foo)) (let* (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-macros-let-with-assignment ()
  "We shouldn't confuse let assignments with macro calls."
  (with-temp-backed-buffer
   "(let ((foo nil) (foo nil))) (let* ((foo nil) (foo nil)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should (null calls)))))

(ert-deftest elisp-refs--find-macros-let-with-assignment-call ()
  "We should find macro calls in let assignments."
  (with-temp-backed-buffer
   "(let ((bar (foo)))) (let* ((bar (foo))))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should
      (equal (length calls) 2)))))

(ert-deftest elisp-refs--find-calls-let-body ()
  "We should find macro calls in let body."
  (with-temp-backed-buffer
   "(let (bar) (foo)) (let* (bar) (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (calls (elisp-refs--read-and-find elisp-refs-buf 'foo
                                            #'elisp-refs--macro-p)))
     (should (equal (length calls) 2)))))

(ert-deftest elisp-refs--find-symbols ()
  "We should find symbols, not their containing forms."
  (with-temp-backed-buffer
   "(foo foo)"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find-symbol elisp-refs-buf 'foo)))
     (should
      (equal
       matches
       (list '(foo 2 5) '(foo 6 9)))))))

(ert-deftest elisp-refs--find-var-basic ()
  "Test the base case of finding variables"
  (with-temp-backed-buffer
   "foo"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should
      (equal
       matches
       (list '(foo 1 4)))))))

(ert-deftest elisp-refs--find-var-in-lambda ()
  "Find variable references in lambda expressions."
  (with-temp-backed-buffer
   "(lambda (foo) (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should
      (equal matches (list (list 'foo 10 13)))))))

(ert-deftest elisp-refs--find-var-ignores-calls ()
  "Function calls are not variable references."
  (with-temp-backed-buffer
   "(baz (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should (null matches)))))

(ert-deftest elisp-refs--find-var-ignores-defs ()
  "Function definitions and macro definitions are not varible references."
  (with-temp-backed-buffer
   "(defun foo ()) (defmacro foo ())"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should (null matches)))))

(ert-deftest elisp-refs--find-var-let-without-assignments ()
  "We should recognise let variables as variable references."
  (with-temp-backed-buffer
   "(let (foo)) (let* (foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should
      (equal
       matches
       (list '(foo 7 10) '(foo 20 23)))))))

(ert-deftest elisp-refs--find-var-let-with-assignments ()
  "We should recognise let variables as variable references."
  (with-temp-backed-buffer
   "(let ((foo 1))) (let* ((foo 2)))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should
      (equal
       matches
       (list '(foo 8 11) '(foo 25 28)))))))

(ert-deftest elisp-refs--find-var-let-body ()
  "We should recognise let variables as variable references."
  (with-temp-backed-buffer
   "(let ((x (1+ foo))) (+ x foo))"
   (let* ((elisp-refs-buf (elisp-refs--contents-buffer (buffer-file-name)))
          (matches (elisp-refs--read-and-find elisp-refs-buf 'foo
                                              #'elisp-refs--variable-p)))
     (should
      (equal
       (length matches)
       2)))))

(ert-deftest elisp-refs--unindent-rigidly ()
  "Ensure we unindent by the right amount."
  ;; Take the smallest amount of indentation, (2 in this case), and
  ;; unindent by that amount.
  (should
   (equal
    (elisp-refs--unindent-rigidly "   foo\n  bar\n    baz")
    " foo\nbar\n  baz"))
  ;; If one of the lines has no indent, do nothing.
  (should
   (equal
    (elisp-refs--unindent-rigidly "foo\n bar")
    "foo\n bar"))
  ;; We should have set elisp-refs-unindented correctly.
  (should
   (equal
    (get-text-property
     0
     'elisp-refs-unindented
     (elisp-refs--unindent-rigidly "  foo"))
    2))
  ;; We should still have elisp-refs-path properties in the entire string.
  (let ((result (elisp-refs--unindent-rigidly
                 (propertize " foo\n bar" 'elisp-refs-path "/foo"))))
    (cl-loop for i from 0 below (length result) do
             (should
              (get-text-property i 'elisp-refs-path result)))))

(ert-deftest elisp-refs--replace-tabs ()
  "Ensure we replace all tabs in STRING."
  (let ((tab-width 4))
    ;; zero tabs
    (should (equal (elisp-refs--replace-tabs " a ") " a "))
    ;; many tabs
    (should (equal (elisp-refs--replace-tabs "a\t\tb") "a        b"))))

(ert-deftest elisp-refs--file-name-handler ()
  "Ensure overriding `file-name-handler-alist' doesn't break our functionality."
  (let ((file-name-handler-alist nil)
        (jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-function 'buffer-substring-no-properties)))

(ert-deftest elisp-refs-function ()
  "Smoke test for `elisp-refs-function'."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-function 'format))
  (should
   (equal (buffer-name) "*refs: format*")))

(ert-deftest elisp-refs-macro ()
  "Smoke test for `elisp-refs-macro'."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-macro 'when))
  (should
   (equal (buffer-name) "*refs: when*")))

(ert-deftest elisp-refs-variable ()
  "Smoke test for `elisp-refs-variable'."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-variable 'case-fold-search))
  (should
   (equal (buffer-name) "*refs: case-fold-search*")))

(ert-deftest elisp-refs-special ()
  "Smoke test for `elisp-refs-special'."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-special 'prog2))
  (should
   (equal (buffer-name) "*refs: prog2*")))

(ert-deftest elisp-refs-symbol ()
  "Smoke test for `elisp-refs-symbol'."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-symbol 'format-message))
  (should
   (equal (buffer-name) "*refs: format-message*")))

(ert-deftest elisp-refs-function-prefix ()
  "Smoke test for searching with a path prefix."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    (elisp-refs-function 'format-message "/usr/share/emacs"))
  (should
   (equal (buffer-name) "*refs: format-message*")))

(ert-deftest elisp-refs-next-match ()
  "Ensure movement commands move between results."
  (let ((jka-compr-verbose nil)
        (elisp-refs-verbose nil))
    ;; First, get a results buffer.
    (elisp-refs-function 'lsh))
  ;; After moving to the first result, we should have a property that
  ;; tells us where the result came from.
  (elisp-refs-next-match)
  (should
   (not (null (get-text-property (point) 'elisp-refs-start-pos)))))


test/test-helper.el
;;; test-helper.el --- Helper for tests              -*- lexical-binding: t; -*-

;; Copyright (C) 2016 Wilfred Hughes

;; Author:  <me@wilfred.me.uk>

;;; Code:

(require 'ert)
(require 'f)

(let ((elisp-refs-dir (f-parent (f-dirname (f-this-file)))))
  (add-to-list 'load-path elisp-refs-dir))

(require 'undercover)
(undercover "elisp-refs.el"
	    (:exclude "*-test.el")
	    (:report-file "/tmp/undercover-report.json"))

(provide 'test-helper)
;;; test-helper.el ends here


callers.el
;;; callers.el --- find callers of elisp functions or macros  -*- lexical-binding: t; -*-

;; Copyright (C) 2016  

;; Author: Wilfred Hughes <me@wilfred.me.uk>
;; Version: 0.1
;; Keywords: lisp
;; Package-Requires: ((dash "2.12.0"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; A package for finding callers of elisp functions or macros. Really
;; useful for finding examples.

;;; Code:

;; `read' can move through a buffer
;; `read-from-string' returns final index.

(require 'dash)

(defun callers--next-sexp-start ()
  "Put point at the start of the next sexp.
Signals 'end if there are no more sexps."
  ;; TODO: write in terms of `scan-sexps'.
  (forward-sexp 1)
  (when (equal (point) (point-max))
    (throw 'after-last-sexp nil))
  (forward-sexp -1))

(defun callers--read-all-forms (buffer)
  "Return a list of all the forms in BUFFER, with the string
indexes where each form starts and ends."
  (let ((forms nil))
    (with-current-buffer buffer
      (catch 'after-last-sexp
        (goto-char (point-min))
        (callers--next-sexp-start)
        ;; Loop until we have no more forms to read.
        (while t
          ;; String indexing is zero-indexed, but point is
          ;; one-indexed.
          (let* ((sexp-start (1- (point)))
                 (form-with-pos (read-from-string (buffer-string) sexp-start))
                 (form (car form-with-pos))
                 (sexp-end (cdr form-with-pos)))
            (push (list form sexp-start sexp-end) forms)
            (goto-char (1+ sexp-end)))
          (callers--next-sexp-start))))
    (nreverse forms)))

(defun callers--find-call (form symbol)
  "If FORM contains a call to SYMBOL, return it.
Returns nil otherwise.

This is basic static analysis, so indirect function calls are
ignored."
  ;; TODO: Handle funcall to static symbols too.
  (cond
   ;; Base case: are we looking at (symbol ...)?
   ((and (consp form) (eq (car form) symbol))
    form)
   ;; Recurse, so we can find (... (symbol ...) ...)
   ((consp form)
    (--any-p (callers--has-call-p it symbol) form))
   ;; If it's not a cons cell, it's not a call.
   (t
    nil)))

(provide 'callers)
;;; callers.el ends here


refs-bench.el
;;; refs-bench.el --- measure refs.el performance

;;; Code:

(require 'refs)
(require 'dash)
(require 'shut-up)

(defmacro refs--print-time (form)
  "Evaluate FORM, and print the time taken."
  `(progn
     (message "Timing %s" ',form)
     (-let [(total-time gc-runs gc-time)
            (shut-up (benchmark-run 1 ,form))]
       (message "Elapsed time: %fs (%fs in %d GCs)"
                total-time
                gc-time
                gc-runs))))

;; TODO: benchmark refs-variable (and add a smoke test)
;; TODO: make this more representative by loading more elisp files
;; before searching. Running this in a GUI is also conspicuously
;; slower, which bench.sh doesn't reflect.
(defun refs-bench ()
  "Measure runtime of searching."
  (interactive)
  (refs--report-loc)
  ;; Measure a fairly uncommon function.
  (refs--print-time (refs-function 'mod))
  ;; Measure a common macro
  (refs--print-time (refs-macro 'when))
  ;; Compare with searching for the same symbol without walking
  (refs--print-time (refs-symbol 'when))
  ;; Synthetic test of a large number of results.
  (message "Formatting 10,000 results")
  (let ((forms (-repeat 10000 (list '(ignored) 1 64)))
        (buf (generate-new-buffer " *dummy-refs-buf*")))
    (with-current-buffer buf
      (insert "(defun foo (bar) (if bar nil (with-current-buffer bar))) ;; blah")
      (setq-local refs--path "/tmp/foo.el"))
    (refs--print-time
     (refs--show-results 'foo "foo bar" (list (cons forms buf))))
    (kill-buffer buf)))

(defun refs--report-loc ()
  "Report the total number of lines of code searched."
  (interactive)
  (let* ((loaded-paths (refs--loaded-files))
         (loaded-src-bufs (-map #'refs--contents-buffer loaded-paths))
         (total-lines (-sum (--map (with-current-buffer it
                                     (line-number-at-pos (point-max)))
                                   loaded-src-bufs))))
    ;; Clean up temporary buffers.
    (--each loaded-src-bufs (kill-buffer it))
    (message "Total LOC: %s" (refs--format-int total-lines))))

(provide 'refs-bench)
;;; refs-bench.el ends here


elisp-refs-bench.el
;;; elisp-refs-bench.el --- measure elisp-refs.el performance

;;; Code:

(require 'elisp-refs)
(require 'dash)

(defmacro elisp-refs--print-time (form)
  "Evaluate FORM, and print the time taken."
  `(progn
     (message "Timing %s" ',form)
     (-let [(total-time gc-runs gc-time)
            (cl-letf (((symbol-function #'message)
                       (lambda (_format-string &rest _args))))
              (benchmark-run 1 ,form))]
       (message "Elapsed time: %fs (%fs in %d GCs)\n"
                total-time
                gc-time
                gc-runs))))

;; TODO: benchmark elisp-refs-variable (and add a smoke test)
;; TODO: make this more representative by loading more elisp files
;; before searching. Running this in a GUI is also conspicuously
;; slower, which bench.sh doesn't reflect.
(defun elisp-refs-bench ()
  "Measure runtime of searching."
  (interactive)
  (elisp-refs--report-loc)
  ;; Measure a fairly uncommon function.
  (elisp-refs--print-time (elisp-refs-function 'mod))
  ;; Measure a common macro
  (elisp-refs--print-time (elisp-refs-macro 'when))
  ;; Compare with searching for the same symbol without walking
  (elisp-refs--print-time (elisp-refs-symbol 'when))
  ;; Synthetic test of a large number of results.
  (message "Formatting 10,000 results")
  (let ((forms (-repeat 10000 (list '(ignored) 1 64)))
        (buf (generate-new-buffer " *dummy-elisp-refs-buf*")))
    (with-current-buffer buf
      (insert "(defun foo (bar) (if bar nil (with-current-buffer bar))) ;; blah")
      (setq-local elisp-refs--path "/tmp/foo.el"))
    (elisp-refs--print-time
     (elisp-refs--show-results 'foo "foo bar" (list (cons forms buf))
                               20 nil))
    (kill-buffer buf)))

(defun elisp-refs--report-loc ()
  "Report the total number of lines of code searched."
  (interactive)
  (let* ((loaded-paths (elisp-refs--loaded-paths))
         (loaded-src-bufs (cl-letf (((symbol-function #'message)
                                     (lambda (_format-string &rest _args))))
                            (-map #'elisp-refs--contents-buffer loaded-paths)))
         (total-lines (-sum (--map (with-current-buffer it
                                     (line-number-at-pos (point-max)))
                                   loaded-src-bufs))))
    ;; Clean up temporary buffers.
    (--each loaded-src-bufs (kill-buffer it))
    (message "Total LOC: %s" (elisp-refs--format-int total-lines))))

(provide 'elisp-refs-bench)
;;; elisp-refs-bench.el ends here
